### PR TITLE
Add sequence type check when using `unordered` with single argument

### PR DESCRIPTION
**Pull request by soxofaan**
Addresses #2

### CODE REVIEW

File: pytest_unordered.py [hunk 1]
```diff
@@ -24,11 +27,14 @@ def __eq__(self, actual: Iterable) -> bool:
         extra_left, extra_right = _compare_eq_unordered(self, actual)
         return not extra_left and not extra_right
 
+    def __ne__(self, actual: Iterable) -> bool:
+        return not self.__eq__(actual)
```
Comment (utapyngo) [inline pytest_unordered.py:31]: Why is this required?

File: pytest_unordered.py [hunk 2]
```diff
@@ -7,14 +7,17 @@
...
+    def __init__(self, expected: Iterable, check_type: bool=True):
         if not isinstance(expected, Iterable):
             raise TypeError(
                 "cannot make unordered comparisons to non-iterable: {!r}".format(expected)
             )
         super().__init__(expected)
+        self._expected_type = type(expected) if check_type else None
 
     def __eq__(self, actual: Iterable) -> bool:
+        if self._expected_type and self._expected_type != type(actual):
```
Comment (utapyngo) [inline pytest_unordered.py:19]:
```suggestion
        if self._expected_type is not None and self._expected_type != type(actual):
```

File: pytest_unordered.py [hunk 3]
```diff
@@ -48,6 +54,11 @@ def pytest_assertrepr_compare(config, op, left, right):
         left_repr = saferepr(left)
         right_repr = saferepr(right)
         result = ["{} {} {}".format(left_repr, op, right_repr)]
+        left_type = left._expected_type if isinstance(left, UnorderedList) else type(left)
+        right_type = right._expected_type if isinstance(right, UnorderedList) else type(right)
+        if left_type and right_type and left_type != right_type:
+            result.append("Type mismatch:")
+            result.append("{l} != {r}".format(l=left_type, r=right_type))
```
Comment (utapyngo) [inline pytest_unordered.py:61]: "l" looks similar to "I" and "1". Please use another name.

File: tests/test_unordered.py [hunk 4]
```diff
@@ -7,20 +7,85 @@
...
+    ["expected", "actual"],
     [
         (unordered(1, 2, 3), [3, 2, 1]),
-        ([3, 2, 1], unordered(1, 2, 3)),
+        (unordered(1, 2, 3), (3, 2, 1)),
+        (unordered(1, 2, 3), {3, 2, 1}),
+        (unordered([1, 2, 3]), [3, 2, 1]),
+        (unordered((1, 2, 3)), (3, 2, 1)),
+        (unordered({1, 2, 3}), {3, 2, 1}),
+        (unordered({1: 2, 3: 4}), {3: 4, 1: 2}),
```
Comment (utapyngo) [inline tests/test_unordered.py:18]: This is ambiguous. What is an unordered `dict`?

File: tests/test_unordered.py [hunk 5]
```diff
@@ -7,20 +7,85 @@
...
+        (unordered("a", "b", "c"), ["b", "a", "c"]),
+        (unordered("a", "b", "c"), "bac"),
     ],
 )
-def test_unordered(left, right):
+def test_unordered(expected, actual):
+    assert expected == actual
+    assert actual == expected
+    with pytest.raises(AssertionError):
+        assert expected != actual
```
Comment (utapyngo) [inline tests/test_unordered.py:34]: Why not use `assert not (expected != actual)` here?

**Review by utapyngo — CHANGES_REQUESTED** [review]

File: pytest_unordered.py (hunk 1 above)
Comment (soxofaan) [reply pytest_unordered.py:31]: If you don't do this, some tests I added fail because unordered "unequality" is not checked properly

I think it is because `UnorderedList` is a subclass of `list`, which has an `__ne__` that does not goes through `__eq__`.

File: pytest_unordered.py (hunk 2 above)
Comment (soxofaan) [reply pytest_unordered.py:19]: ok will do

File: pytest_unordered.py (hunk 3 above)
Comment (soxofaan) [reply pytest_unordered.py:61]: good point

File: tests/test_unordered.py (hunk 4 above)
Comment (soxofaan) [reply tests/test_unordered.py:18]: I added this because `Unordered`  expects an `Iterable` and as `dict` is also an `Iterable` I added this less obvious example. In practice it indeed does not make a lot of sense. You want to drop it?

File: tests/test_unordered.py (hunk 5 above)
Comment (soxofaan) [reply tests/test_unordered.py:34]: I wanted to check the assertion message through `raises( ... match=...)` but that turned out not very interesting. I can indeed do it with `assert not`

File: tests/test_unordered.py (hunk 4 above)
Comment (utapyngo) [reply tests/test_unordered.py:18]: I don't want to drop this line, but I want to present least surprise to the users. Currently, `unordered({1: 2, 3: 4}) == {3: 5, 1: 6}` will also hold, which the users don't expect.

Comment (soxofaan) [reply tests/test_unordered.py:18]: oh yes, I didn't think of that. Maybe it's best to reject mapping types in unordered then to avoid accidental surprise?

Comment (utapyngo) [reply tests/test_unordered.py:18]: It is a good idea. I was thinking about keeping the old behaviour for dicts, i.e. `unordered({1: 2, 3: 4}) == [{1: 2, 3: 4}]`, but this would not be in line with your changes.

File: pytest_unordered.py (hunk 1 above)
Comment (utapyngo) [reply pytest_unordered.py:31]: Will the tests pass with this implementation?

```suggestion
    def __ne__(self, actual: Iterable) -> bool:
        return not self == actual
```

Comment (soxofaan) [reply pytest_unordered.py:31]: yep that works too, I'll change it to that

File: tests/test_unordered.py (hunk 4 above)
Comment (soxofaan) [reply tests/test_unordered.py:18]: > unordered({1: 2, 3: 4}) == [{1: 2, 3: 4}]

this looks weird to me to be honest, I guess you see `unordered` as a factory for a new object (a list that does not care about order), while I see it more as a wrapper around an existing sequence (to ignore order when checking equality).

What do you propose to resolve this? Do I remove the test case to finish up this PR and leave the dict handling to another issue?

Comment (utapyngo) [reply tests/test_unordered.py:18]: It should be resolved here. Please implement your idea: rejecting mapping types in `unordered`.

Comment (soxofaan) [reply tests/test_unordered.py]: ok done

**Review by utapyngo — APPROVED** [review]

File: tests/test_unordered.py (hunk 4 above)
Comment (utapyngo) [reply tests/test_unordered.py]: Thank you!

**Comment by utapyngo** [comment]
Thank you, I like the changes and the tests.

Do you think we need to add an AUTHORS file?

**Review by Mogost — APPROVED** [review]

**Comment by soxofaan** [comment]
> Do you think we need to add an AUTHORS file?

that can be done separately from this PR I guess?

**Comment by utapyngo** [comment]
The coverage has dropped for some reason...

File: pytest_unordered.py [hunk 6]
```diff
@@ -1,20 +1,27 @@
...
+                "cannot make unordered comparisons to mapping: {!r}".format(expected)
+            )
         super().__init__(expected)
+        self._expected_type = type(expected) if check_type else None
 
     def __eq__(self, actual: Iterable) -> bool:
+        if self._expected_type is not None and self._expected_type != type(actual):
+            return False
         if not isinstance(actual, Iterable):
             return self.copy() == actual
```
Comment (utapyngo) [inline pytest_unordered.py:26]: After adding the type check, the condition `isinstance(actual, Iterable)` is always `False`.

Comment (soxofaan) [reply pytest_unordered.py:26]: something like this visits those lines:

        assert unordered(1, 2, 3) == 3

Comment (soxofaan) [reply pytest_unordered.py:26]: see PR #4
